File: neetcode150/06_linked_list/0146_lru_cache.cpp
/*
 * =====================================================================
 * LeetCode 146 : LRU Cache                                     (Medium)
 * https://leetcode.com/problems/lru-cache/
 * Category   : Linked List
 * ---------------------------------------------------------------------
 * PROBLEM
 *   Design a cache with capacity C supporting O(1) average:
 *     get(key) → value, or −1 if absent. Access marks key RECENT.
 *     put(k,v) → insert (or overwrite); if full, EVICT the
 *                LEAST RECENTLY USED entry first.
 * ---------------------------------------------------------------------
 * INTUITION
 *   A hash map alone gives O(1) get/put but cannot tell you "which
 *   key was least recently used" — that needs an ORDER, and the
 *   order must be updatable in O(1) whenever a key is touched.
 *   A doubly linked list (LRU ⇔ MRU) delivers: moving a node to the
 *   MRU end costs O(1) with the right pointers. Map key → list node.
 *   Map = the O(1) lookup; list = the O(1) eviction order.
 * ---------------------------------------------------------------------
 * APPROACH 1 — Hash map key → node + doubly linked list (BEST)
 *   DLL keeps keys in recency order: head = LRU (eviction victim),
 *   tail = MRU. get(k): map hit → move node to tail → return value.
 *   put(k,v): exists → update + move to tail; new → append at tail,
 *   and if size > C, unlink head and erase it from the map.
 *   Time  Complexity : O(1) average per get/put.
 *   Space Complexity : O(C) — the map + the list hold ≤ C entries.
 * ---------------------------------------------------------------------
 * APPROACH 2 — Naive: map key → (value, lastUsedTick)
 *   Every get/put stamps the current tick counter. Eviction: scan
 *   all entries for the smallest tick — O(C) per eviction.
 *   Time  Complexity : O(1) get, O(C) put-when-full.
 *   Space Complexity : O(C).
 * ---------------------------------------------------------------------
 * DEEP DIVE — hash map (probing) + doubly linked list marriage
 *   Python dict keys use open addressing with LINEAR PROBING:
 *       h(k, i) = (hash(k) + i) mod m ,  i = 0, 1, 2, …
 *   C++ std::unordered_map uses separate chaining. Both give the
 *   O(1) average random access the DLL cannot provide alone.
 *   The DLL is the OPPOSITE structure: O(1) insert/delete at known
 *   positions but O(n) search — together they cover each other's
 *   weakness: map finds the node, list reorders it. This exact pair
 *   (hash table + doubly linked list) is how real caches work:
 *   Redis' allkeys-lru, memcached's LRU, and the Linux kernel's
 *   page cache (radix tree + LRU list). Eviction = "drop the head",
 *   promotion = "move to tail" — every browser tab, every CDN.
 * =====================================================================
 */

#include <unordered_map>   // std::unordered_map — O(1) key lookup
#include <vector>          // std::vector — op sequences for tests
#include <chrono>          // std::chrono — timing
#include <iostream>        // std::cout — output
#include <sys/resource.h>  // getrusage — peak RAM
#include "list_node.h"     // our header (ListNode unused here; the
                           // lesson: headers centralize shared types)

/* =====================================================================
 * APPROACH 1 : Hash map + custom doubly linked list (BEST)
 * ===================================================================== */
class LRUCache_List {
private:
    // Doubly linked list node: key, value, prev, next.
    struct DNode {
        int key;
        int value;
        DNode* prev;
        DNode* next;
        DNode(int k, int v) : key(k), value(v), prev(nullptr), next(nullptr) {}
    };

    int capacity_;
    // map: key → pointer to its list node (O(1) location).
    std::unordered_map<int, DNode*> map_;
    DNode* head_;   // LRU end — eviction victim lives here
    DNode* tail_;   // MRU end — most recently used lives here

    // Move an existing node to the MRU end (tail). Four pointer
    // rewrites on a doubly linked list — O(1).
    void moveToTail(DNode* node) {
        if (node == tail_) {
            return;   // already most recent
        }
        // Unlink: bridge prev <-> next around the node.
        if (node->prev != nullptr) node->prev->next = node->next;
        if (node->next != nullptr) node->next->prev = node->prev;
        if (node == head_) head_ = node->next;   // node was the victim
        // Re-attach at the tail.
        node->prev = tail_;
        node->next = nullptr;
        tail_->next = node;
        tail_ = node;
    }

    // Append a BRAND-NEW node at the MRU end.
    void pushBack(int key, int value) {
        DNode* node = new DNode(key, value);
        map_[key] = node;
        if (tail_ == nullptr) {
            head_ = tail_ = node;   // first node ever
        } else {
            node->prev = tail_;
            tail_->next = node;
            tail_ = node;
        }
    }

    // Evict the LRU (head) — O(1).
    void popFront() {
        DNode* victim = head_;
        head_ = head_->next;
        if (head_ != nullptr) {
            head_->prev = nullptr;
        } else {
            tail_ = nullptr;
        }
        map_.erase(victim->key);   // forget the key too
        delete victim;
    }

public:
    /*
     * Purpose : Construct a cache with the given capacity.
     * Inputs  : capacity — maximum number of entries before eviction.
     */
    LRUCache_List(int capacity) : capacity_(capacity), head_(nullptr), tail_(nullptr) {}

    /*
     * Purpose : Return the value of `key` if cached, else −1.
     * Inputs  : key — the lookup key.
     * Output  : The cached value, or −1. A hit refreshes recency.
     */
    int get(int key) {
        auto it = map_.find(key);
        if (it == map_.end()) {
            return -1;   // miss: nothing to refresh
        }
        moveToTail(it->second);   // hit ⇒ now the MOST recent
        return it->second->value;
    }

    /*
     * Purpose : Insert or overwrite key→value; evict LRU if full.
     * Inputs  : key — the key to store; value — the value to store.
     * Output  : Returns nothing (void).
     */
    void put(int key, int value) {
        auto it = map_.find(key);
        if (it != map_.end()) {
            // Existing key: update the value, refresh recency.
            it->second->value = value;
            moveToTail(it->second);
            return;
        }
        // New key: if full, evict the least-recently-used first.
        if (static_cast<int>(map_.size()) == capacity_) {
            popFront();
        }
        pushBack(key, value);
    }
};

/* =====================================================================
 * APPROACH 2 : Naive tick counter + linear eviction scan (contrast)
 * ===================================================================== */
class LRUCache_Naive {
private:
    int capacity_;
    long long tick_;   // global clock, incremented on every op
    // key → (value, lastUsedTick)
    std::unordered_map<int, std::pair<int, long long>> store_;

public:
    /*
     * Purpose : Construct a cache with the given capacity.
     * Inputs  : capacity — maximum number of entries before eviction.
     */
    LRUCache_Naive(int capacity) : capacity_(capacity), tick_(0) {}

    /*
     * Purpose : Return the value of `key` if cached, else −1.
     * Inputs  : key — the lookup key.
     * Output  : The cached value, or −1. A hit refreshes recency.
     */
    int get(int key) {
        auto it = store_.find(key);
        if (it == store_.end()) {
            return -1;
        }
        it->second.second = ++tick_;   // stamp the access time
        return it->second.first;
    }

    /*
     * Purpose : Insert or overwrite key→value; evict LRU if full.
     * Inputs  : key — the key to store; value — the value to store.
     * Output  : Returns nothing (void).
     */
    void put(int key, int value) {
        auto it = store_.find(key);
        if (it != store_.end()) {
            it->second = {value, ++tick_};
            return;
        }
        if (static_cast<int>(store_.size()) == capacity_) {
            // Eviction needs the OLDEST stamp — scan everything.
            int victimKey = -1;
            long long oldest = tick_;
            for (const auto& entry : store_) {
                if (entry.second.second < oldest) {
                    oldest = entry.second.second;
                    victimKey = entry.first;
                }
            }
            store_.erase(victimKey);
        }
        store_[key] = {value, ++tick_};
    }
};

/* =====================================================================
 * BENCHMARK — official example + a burst, time + peak RAM
 * ===================================================================== */
int main() {
    // Official example: capacity 2,
    // put(1,1) put(2,2) get(1) put(3,3) get(2) put(4,4) get(1) get(3) get(4)
    // → [null, null, 1, null, -1, null, -1, 3, 4]
    std::vector<int> expected = {1, -1, -1, 3, 4};

    auto runExample = [&](auto& cache) {
        std::vector<int> out;
        cache.put(1, 1);
        cache.put(2, 2);
        out.push_back(cache.get(1));   // 1  (2 was least recent)
        cache.put(3, 3);               // evicts 2
        out.push_back(cache.get(2));   // -1 (gone)
        cache.put(4, 4);               // evicts 1
        out.push_back(cache.get(1));   // -1 (gone)
        out.push_back(cache.get(3));   // 3
        out.push_back(cache.get(4));   // 4
        return out == expected;
    };

    bool ok1 = false, ok2 = false;

    // ---- Approach 1 ----
    struct rusage r1a, r1b;
    getrusage(RUSAGE_SELF, &r1a);
    auto s1 = std::chrono::high_resolution_clock::now();
    {
        LRUCache_List cache(2);
        ok1 = runExample(cache);
    }
    auto e1 = std::chrono::high_resolution_clock::now();
    getrusage(RUSAGE_SELF, &r1b);

    // ---- Approach 2 ----
    struct rusage r2a, r2b;
    getrusage(RUSAGE_SELF, &r2a);
    auto s2 = std::chrono::high_resolution_clock::now();
    {
        LRUCache_Naive cache(2);
        ok2 = runExample(cache);
    }
    auto e2 = std::chrono::high_resolution_clock::now();
    getrusage(RUSAGE_SELF, &r2b);

    double us1 = std::chrono::duration_cast<std::chrono::nanoseconds>(e1 - s1).count() / 1000.0;
    double us2 = std::chrono::duration_cast<std::chrono::nanoseconds>(e2 - s2).count() / 1000.0;
    long kb1 = (r1b.ru_maxrss - r1a.ru_maxrss) / 1024;
    long kb2 = (r2b.ru_maxrss - r2a.ru_maxrss) / 1024;

    std::cout << "0146 LRU Cache\n";
    std::cout << "Approach 1 (map + DLL) : time = " << us1 << " µs  peak-ram-delta = " << kb1 << " KB\n";
    std::cout << "Approach 2 (naive tick) : time = " << us2 << " µs  peak-ram-delta = " << kb2 << " KB\n";
    std::cout << (ok1 && ok2 ? "PASS : both caches reproduced the official example (1, -1, -1, 3, 4)."
                             : "FAIL : at least one cache deviated from the expected outputs.")
              << "\n";
    return 0;
}

File: neetcode150/06_linked_list/0146_lru_cache.py
"""
=====================================================================
LeetCode 146 : LRU Cache                                     (Medium)
https://leetcode.com/problems/lru-cache/
Category   : Linked List
---------------------------------------------------------------------
PROBLEM
    Design a cache with capacity C supporting O(1) average:
      get(key) → value, or −1 if absent. Access marks key RECENT.
      put(k,v) → insert/overwrite; if full, EVICT the LEAST
                 RECENTLY USED entry first.
---------------------------------------------------------------------
INTUITION
    A hash map alone gives O(1) get/put but no "which key is least
    recently used?" — that needs an ORDER updatable in O(1). A
    doubly linked list (LRU ⇔ MRU) delivers: moving a node to the
    MRU end is O(1). Map key → list node. Two structures, one job.
---------------------------------------------------------------------
APPROACH 1 — Hash map key → node + doubly linked list (BEST)
    DLL keeps recency order: head = LRU (victim), tail = MRU.
    get: hit → move node to tail. put: exists → update + move to
    tail; new → append at tail; if full → unlink head + erase map.
    Time  Complexity : O(1) average per op   Space : O(C)
---------------------------------------------------------------------
APPROACH 2 — Naive: key → (value, lastUsedTick)
    Every op stamps a global tick. Eviction scans ALL entries for
    the smallest tick — O(C) per eviction.
    Time  Complexity : O(1) get, O(C) put-when-full   Space : O(C)
---------------------------------------------------------------------
REAL LIFE
    • Browser back/forward + tab caches, CDN edge caches, Redis
      allkeys-lru, memcached, the Linux page cache — all are the
      hash-map + doubly-linked-list marriage.
    • LRU = the default cache policy because locality of reference
      is a law of real workloads.
=====================================================================
"""

from __future__ import annotations
import time
import tracemalloc
from typing import Optional


class _DNode:
    """Doubly linked list node: key, value, prev, next."""

    __slots__ = ("key", "value", "prev", "next")

    def __init__(self, key: int = 0, value: int = 0) -> None:
        self.key = key
        self.value = value
        self.prev: Optional[_DNode] = None
        self.next: Optional[_DNode] = None


# =====================================================================
# APPROACH 1 : Hash map + doubly linked list (best)
# =====================================================================
class LRUCache_List:
    """
    Purpose : An LRU cache with O(1) average get/put.
    Inputs  : capacity — maximum entries before eviction.
    Output  : get returns the value or −1; put stores/evicts in place.
    """

    def __init__(self, capacity: int) -> None:
        self.capacity = capacity
        self.map: dict[int, _DNode] = {}    # key → list node (O(1) find)
        self.head: Optional[_DNode] = None  # LRU end — victim lives here
        self.tail: Optional[_DNode] = None  # MRU end — most recent here

    def _move_to_tail(self, node: _DNode) -> None:
        """Refresh recency: unlink node, re-attach at the MRU end."""
        if node is self.tail:
            return                          # already most recent
        # Unlink: bridge prev <-> next around the node.
        if node.prev is not None:
            node.prev.next = node.next
        if node.next is not None:
            node.next.prev = node.prev
        if node is self.head:
            self.head = node.next           # node was the victim end
        # Re-attach at the tail.
        node.prev = self.tail
        node.next = None
        if self.tail is not None:
            self.tail.next = node
        self.tail = node

    def _push_back(self, key: int, value: int) -> None:
        """Append a brand-new node at the MRU end."""
        node = _DNode(key, value)
        self.map[key] = node
        if self.tail is None:
            self.head = self.tail = node    # first node ever
        else:
            node.prev = self.tail
            self.tail.next = node
            self.tail = node

    def _pop_front(self) -> None:
        """Evict the LRU (head) — O(1)."""
        victim = self.head
        if victim is None:
            return
        self.head = victim.next
        if self.head is not None:
            self.head.prev = None
        else:
            self.tail = None
        del self.map[victim.key]            # forget the key too

    def get(self, key: int) -> int:
        """Return the value of key, or −1; a hit refreshes recency."""
        node = self.map.get(key)
        if node is None:
            return -1
        self._move_to_tail(node)            # hit ⇒ now the MOST recent
        return node.value

    def put(self, key: int, value: int) -> None:
        """Insert/overwrite key→value; evict LRU if the cache is full."""
        node = self.map.get(key)
        if node is not None:
            node.value = value              # overwrite in place
            self._move_to_tail(node)
            return
        if len(self.map) == self.capacity:
            self._pop_front()               # full → evict least recent
        self._push_back(key, value)


# =====================================================================
# APPROACH 2 : Naive tick counter + linear eviction scan (contrast)
# =====================================================================
class LRUCache_Naive:
    """
    Purpose : An LRU cache using timestamps and a linear eviction scan.
    Inputs  : capacity — maximum entries before eviction.
    Output  : get returns the value or −1; put stores/evicts in place.
    Note    : get is O(1); put can be O(C) — the contrast to A1.
    """

    def __init__(self, capacity: int) -> None:
        self.capacity = capacity
        self.tick = 0
        self.store: dict[int, tuple[int, int]] = {}   # key → (value, tick)

    def get(self, key: int) -> int:
        """Return the value of key, or −1; a hit refreshes recency."""
        entry = self.store.get(key)
        if entry is None:
            return -1
        self.tick += 1
        self.store[key] = (entry[0], self.tick)       # stamp access time
        return entry[0]

    def put(self, key: int, value: int) -> None:
        """Insert/overwrite key→value; evict LRU if the cache is full."""
        if key in self.store:
            self.tick += 1
            self.store[key] = (value, self.tick)
            return
        if len(self.store) == self.capacity:
            # Eviction needs the OLDEST stamp — scan everything (O(C)).
            victim_key = min(self.store, key=lambda k: self.store[k][1])
            del self.store[victim_key]
        self.tick += 1
        self.store[key] = (value, self.tick)


# =====================================================================
# BENCHMARK — official example, both approaches, time + peak memory
# =====================================================================
if __name__ == "__main__":
    # Official example: cap 2, ops
    # put(1,1) put(2,2) get(1) put(3,3) get(2) put(4,4) get(1) get(3) get(4)
    # → [1, -1, -1, 3, 4]
    expected = [1, -1, -1, 3, 4]

    def run_example(cache) -> bool:
        out: list[int] = []
        cache.put(1, 1)
        cache.put(2, 2)
        out.append(cache.get(1))    # 1  (2 is now least recent)
        cache.put(3, 3)             # evicts 2
        out.append(cache.get(2))    # -1 (gone)
        cache.put(4, 4)             # evicts 1
        out.append(cache.get(1))    # -1 (gone)
        out.append(cache.get(3))    # 3
        out.append(cache.get(4))    # 4
        return out == expected

    # --- Approach 1 ---
    tracemalloc.start()
    t0 = time.perf_counter()
    ok1 = run_example(LRUCache_List(2))
    t1 = time.perf_counter()
    _, peak1 = tracemalloc.get_traced_memory()
    tracemalloc.stop()

    # --- Approach 2 ---
    tracemalloc.start()
    t0b = time.perf_counter()
    ok2 = run_example(LRUCache_Naive(2))
    t1b = time.perf_counter()
    _, peak2 = tracemalloc.get_traced_memory()
    tracemalloc.stop()

    print("0146 LRU Cache")
    print(f"Approach 1 (map + DLL)  : time = {(t1 - t0) * 1e6:.2f} µs  peak-mem = {peak1} bytes")
    print(f"Approach 2 (naive tick) : time = {(t1b - t0b) * 1e6:.2f} µs  peak-mem = {peak2} bytes")
    print("PASS : both caches reproduced the official example (1, -1, -1, 3, 4)."
          if ok1 and ok2 else "FAIL : at least one cache deviated from the expected outputs.")
